Keeping release history

stephenharris opened the pull request:
This allows the release history to be kept (and would close #3). Please note that I've not tested multiple users trying to release to the same repo, and thus how it handles merge conflicts (probably will just error out). 

The idea is to clone the release repository to a temporary directory (I think this should probably be configurable), copy the source directory into that temporary directory, commit the changes and push.

I had to update grunt-contrib-nodeunit in order for the tests to pass (which they now do).

I've also got a second pull-request for tagging a release commit, but its branched from this feature branch.

stephenharris commented:
@iclanzan - sorry to pest, but any interest in merging this? If not I'll create a fork, as currently I have a composer file with `dev-` in it that I want rid of :).

iclanzan commented on tasks/git_deploy.js:
```diff
@@ -35,22 +39,54 @@ module.exports = function(grunt) {
       };
     }
 
+    function copyDirectory( src_dir, dest_dir ){
```
Why not use camel case for identifiers?

stephenharris replied:
Habit. Fixed in 7b41d27

iclanzan commented on tasks/git_deploy.js:
```diff
@@ -35,22 +39,54 @@ module.exports = function(grunt) {
+      copyDirectory( src, deployDir ),
       git(['add', '--all']),
       git(['commit', '--message="' + options.message + '"']),
       git(['push', '--prune', '--force', '--quiet', options.url, options.branch])
```
Is `force` still needed?

stephenharris replied:
Depends - I think `--force` keeps existing behaviour in that if you encounter a merge conflict it will override the history.

iclanzan commented:
Can you please add unit tests for the feature you are introducing?

stephenharris commented:
Ok so in the course of unit testing I've come across an issue which I hadn't encountered while doing exploratory testing with a 'live' repo: `git clone` pulls in the master branch. On the second deploy this means the history of the deployment branch is not being pulled in (unless it happens to `master`). (The fact we were using `--force` disguised this fact). Specifying the branch when cloning solves this problem.

This raised another problem: what if the deployment branch doesn't exist? (which it doesn't in the unit tests). This is not likely to be a problem in general use, so working with the assumption that it does indeed exist, in the tests I commit a file to the `repo` repository on the `gh-pages` branch immediately after setting the repository up. 

updates and unit tests to follow....

stephenharris commented:
I've just fixed a few more bugs that I encountered:

 - Deleted files were not deleted from the release repository. (You clone the repo and then copy the source directory into, I had forgotten the key step of removing existing files before copying)
- Dot files were ignored because you need to explicitly include them with `dot:true`

Further tests were added for each of these, as well as for sub-directories.

iclanzan commented:
Is this ready to merge?

stephenharris commented:
Yes, if you're happy with it.

iclanzan commented:
Good job on this!

stephenharris commented:
Thanks for the merge :). I'll be opening another shortly to expose an option to automatically tag releases
